docs/src/dev/future/proposal-scoping-5.asciidoc

```diff
@@ -0,0 +1,157 @@
+
+===== Proposal 1: By default, the Gremlin semantics shall NOT prescribe lazy vs. eager evaluation order =====
+Of course, this does not prevent implementations from opting into a specific evaluation order (the Apache Tinkerpop OLTP processor, for instance, would likely continue to implement a lazy evaluation paradigm and hence may provide more specific guarantees than what is prescribed by Gremlin as a query language). Concretely, the required changes for TP4 in this regard would be to update the documentation, to be explicit about the fact that Gremlin as a language does neither prescribe eager nor lazy evaluation order in the general case, and review (and, where necessary, relax) some existing test cases to be less constraining when it comes to enforcing lazy evaluation.
+
+
+[code]
+----
+Example:
+========
+
+# In the absence of lazy vs. eager evaluation guarantees as proposed above, the  
+# sample query from the Introduction may return different results, depending on
+# the control flow strategy chosen by a specific Gremlin processor:
+gremlin> g.V().hasLabel('person').groupCount('x').select('x')
+
+## Sample result 1:
+# An implementation that internally implements a lazy execution approach may
+# choose to execute traversers sequentially and return the following result
+# (this is the result returned by the Tinkerpop OLTP processor today):
+==>[v[1]:1]
+==>[v[1]:1,v[2]:1]
+==>[v[1]:1,v[2]:1,v[4]:1]
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1]
+
+## Sample result 2:
+# An implementation that internally implements an eager execution approach may
+# choose to batch process results and would return the following result:
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1]
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1]
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1]
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1]
+
+## Sample result 3:
+# Implementations are also free to do vectored processing, e.g. implement "partial 
+# batching" of the results, in which case the following result might be observed:
+==>[v[1]:1,v[2]:1]
+==>[v[1]:1,v[2]:1]
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1] 
+==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1] 
+----
+
+
+===== Proposal 2: The recipe to achieve lazy evaluation is to wrap the relevant part of the query into a local() step =====
+This already works today and could be documented as a __general pattern__ to enforce lazy evaluation for certain parts of the query.
+
+[code]
+----
+Example:
+========
+
+# By wrapping the groupCount() and select() into a local() step, users can enforce lazy
+# execution behavior:
+gremlin> g.V().hasLabel('person').local(groupCount('x').select('x'))
```

Comment:
I don't believe `local()` as it currently exists is a perfect tool to enforce lazy evaluation. When running in TinkerGraph, it does indeed produce the intended result for this example, however it can break down when dealing with bulked traversers. Local step processes one (possibly bulked) traverser at a time which allows it to sometimes operate on multiple values at once.

With LazyBarrierStrategy disabled (to avoid hidden barrier() steps), the following example works as expected with a lazy evaluation:
```
﻿gremlin> g.withoutStrategies(LazyBarrierStrategy).V().both().hasLabel('person').local(groupCount('x').select('x'))
==>[v[2]:1]
==>[v[2]:1,v[4]:1]
==>[v[1]:1,v[2]:1,v[4]:1]
==>[v[1]:2,v[2]:1,v[4]:1]
==>[v[1]:2,v[2]:1,v[4]:2]
==>[v[1]:2,v[2]:1,v[4]:2,v[6]:1]
==>[v[1]:3,v[2]:1,v[4]:2,v[6]:1]
==>[v[1]:3,v[2]:1,v[4]:3,v[6]:1]
```

However, if a barrier is injected prior to the `local()` step, the result is a mix of lazy and eager evaluation:
```
gremlin> g.withoutStrategies(LazyBarrierStrategy).V().both().hasLabel('person').barrier().local(groupCount('x').select('x'))
==>[v[2]:1]
==>[v[2]:1,v[4]:3]
==>[v[2]:1,v[4]:3]
==>[v[2]:1,v[4]:3]
==>[v[1]:3,v[2]:1,v[4]:3]
==>[v[1]:3,v[2]:1,v[4]:3]
==>[v[1]:3,v[2]:1,v[4]:3]
==>[v[1]:3,v[2]:1,v[4]:3,v[6]:1]
```

Unfortunately `flatMap()` also does not produce the intended results in this case as it pushes a single value from a bulked traverser through the child traversal, and then reapply the bulk to the result, instead of processing each value individually:
```
gremlin> g.withoutStrategies(LazyBarrierStrategy).V().both().hasLabel('person').barrier().flatMap(groupCount('x').select('x'))
==>[v[2]:1]
==>[v[2]:1,v[4]:1]
==>[v[2]:1,v[4]:1]
==>[v[2]:1,v[4]:1]
==>[v[1]:1,v[2]:1,v[4]:1]
==>[v[1]:1,v[2]:1,v[4]:1]
==>[v[1]:1,v[2]:1,v[4]:1]
==>[v[1]:1,v[2]:1,v[4]:1,v[6]:1]
```

I'm not aware of any steps in TinkerPop that address this specific concern as I don't believe there have been significant attempts to control evaluation ordering in the past. In order to produce the intended behaviour here, we will need a step which works similarly to `flatMap()`, but instead actually executes the child traversal on each individual value instead of applying a bulk to the result. This would be inefficient and unhelpful for all cases except those which have some sort of aggregation in the child traversal.